crates/ruff_linter/src/checkers/ast/mod.rs

```diff
@@ -3107,18 +3105,33 @@ pub(crate) fn check_ast(
 /// a [`Violation`] to the contained [`OldDiagnostic`] collection on `Drop`.
 pub(crate) struct LintContext<'a> {
     diagnostics: RefCell<Vec<OldDiagnostic>>,
-    source_file: &'a SourceFile,
-    settings: &'a LinterSettings,
+    source_file: SourceFile,
+    rules: Cow<'a, RuleTable>,
 }
 
 impl<'a> LintContext<'a> {
     /// Create a new collector with the given `source_file` and an empty collection of
     /// `OldDiagnostic`s.
-    pub(crate) fn new(source_file: &'a SourceFile, settings: &'a LinterSettings) -> Self {
+    pub(crate) fn new(path: &Path, contents: &str, settings: &'a LinterSettings) -> Self {
+        let source_file =
+            SourceFileBuilder::new(path.to_string_lossy().as_ref(), contents).finish();
+
+        // Ignore diagnostics based on per-file-ignores.
+        let rules = if settings.per_file_ignores.is_empty() {
+            Cow::Borrowed(&settings.rules)
```

Comment:
I wonder if it's cheaper to just clone the rule table. It's just a large array or we could consider using an `Rc` instead